pkg/controller/machine/machine.go

```diff
@@ -888,8 +911,31 @@ func (r *Reconciler) getNode(instance instance.Instance, provider providerconfig
 				return node.DeepCopy(), true, nil
 			}
 		}
+		// If we were unable to find Node by ProviderID, fallback to IP address matching.
+		// This usually happens if there's no CCM deployed in the cluster.
+		//
+		// This mechanism is not always reliable, as providers reuse the IP addresses after
+		// some time.
+		//
+		// If we rollout a Machine, it can happen that a new instance has the same
+		// IP addresses as the instance that has just been deleted. If machine-controller
+		// processes the new Machine before removing the old Machine and the corresponding
+		// Node object, machine-controller could update the NodeOwner label on the old Node
+		// object to point to the Machine that just got created, as IP addresses would match.
+		// This causes machine-controller to fail to delete the old Node object, which could
+		// then cause cluster stability issues in some cases.
 		for _, nodeAddress := range node.Status.Addresses {
 			for instanceAddress := range instance.Addresses() {
+				// Hetzner is known to assign old IP addresses to new instances right after
+				// the old instance has been deleted. On Hetzner, the described issue happens much
```

Comment:
> On Hetzner, the described issue happens much more often than for other providers

I don't think we have evidences that it happens more often with Hetzner than with other providers we support, simply we observed it with Hetzner. I think that almost all clusters we use for manual or automated tests have a CCM deployed. That said it's a picky comment, I let you decide if you want to rectify or keep it ;-)

Reply:
```
// We observed that the issue described above happens often on Hetzner.
// As we know that the Node and the instance name will always be same
// on Hetzner, we can use it as an additional check to prevent this
// issue.
// TODO: We should do this for other providers, but there are providers where
// the node and the instance names will not match, so it requires further
// investigation (e.g. AWS).
```

@irozzo-1A What do you think about the comment above?

Reply:
@irozzo-1A Pushed the comment update.